### PR TITLE
Feature/ntt 320 add filtering and sorting for products list

### PR DESCRIPTION
**Pull request by IhorMurashko**

### CODE REVIEW

File: src/main/java/com/challengeteam/shop/web/controller/PhoneFilteringController.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    @GetMapping("/by")
+    public ResponseEntity<PageResponseDto<PhoneResponseDto>>
+    filterPhones(@Parameter(description = "Filter criteria for phones", required = true)
+                 @Valid @ModelAttribute PhoneMultipleFilterRequest filterRequest,
+                 @Parameter(description = "Page number (starts from 1)", example = "1")
+                 @RequestParam(defaultValue = "1") @Min(1) int page,
+                 @Parameter(description = "Number of items per page", example = "12")
+                 @RequestParam(defaultValue = "12") @Min(1) int size) {
+        log.debug("filterPhones: filterRequest = {}, page = {}, size = {}", filterRequest, page, size);
+        Pageable pageable = PageRequest.of(page - 1, size);
```
Comment (mykola-yefremov) [inline src/main/java/com/challengeteam/shop/web/controller/PhoneFilteringController.java:61]: Can we validate minPrice <= maxPrice here before calling the service? AdminProductController already returns 400 for this case, but this public endpoint currently accepts minPrice > maxPrice and just returns an empty page. Please add the same validation plus an integration test for the reversed range.

File: src/main/java/com/challengeteam/shop/service/pagination/PhoneFilteringAndSortingServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                .add(filterDto.brands(), brands -> (root, query, builder)
+                        -> root.get(Phone_.BRAND).in(brands))
+                .add(filterDto.minPrice(), minPrice -> (root, query, builder)
+                        -> builder.greaterThanOrEqualTo(root.get(Phone_.PRICE), minPrice))
+                .add(filterDto.maxPrice(), maxPrice -> (root, query, builder)
+                        -> builder.lessThanOrEqualTo(root.get(Phone_.PRICE), maxPrice))
+                .add(filterDto.inStock(), () -> (root, query, builder)
+                        -> builder.greaterThan(root.get(Phone_.STOCK), 0))
+                .add(filterDto.preOrder(), () -> (root, query, builder)
+                        -> builder.equal(root.get(Phone_.status), ProductStatus.OUT_OF_STOCK))
```
Comment (mykola-yefremov) [inline src/main/java/com/challengeteam/shop/service/pagination/PhoneFilteringAndSortingServiceImpl.java]: The request/tests define preOrder=true as “stock = 0”, but here we filter by status OUT_OF_STOCK. These are not guaranteed to be the same invariant. Could we either rename this filter to outOfStock/status-based filtering, or change the predicate to stock == 0 and add a test where stock and status differ?

File: src/main/java/com/challengeteam/shop/config/security/SecurityConfig.java [hunk 3]
```diff
@@ -66,6 +66,7 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http,
                         .requestMatchers("/docs/**", "/swagger-ui/**").permitAll()
                         .requestMatchers("/api/v1/test-data/**").permitAll()
                         .requestMatchers("/api/v1/admin/**").hasRole("ADMIN")
+                        .requestMatchers("api/v1/filter/**").permitAll()
```
Comment (mykola-yefremov) [inline src/main/java/com/challengeteam/shop/config/security/SecurityConfig.java:69]: Please add a no-auth integration test for this endpoint. The controller tests still pass Authorization headers everywhere, so the new permitAll rule is not actually verified. Also worth making the matcher consistent with the other rules (`/api/v1/filter/**`).

**Review by mykola-yefremov — CHANGES_REQUESTED** [review]
Nice work overall, especially on covering the main filtering/sorting flows. I am requesting changes because there are a few edge cases that can lead to inconsistent behavior: reversed price ranges are accepted, preOrder semantics dont match the tests/DTO docs, and the new permitAll rule is not verified without Authorization.

Please address the inline comments and add the missing tests. After that I am good with this PR.

**Review by mykola-yefremov — APPROVED** [review]
Thanks for addressing the comments. The reversed price range validation is now covered, preOrder semantics are aligned with stock-based filtering, and the anonymous access path has an integration test.
